### PR TITLE
Generic UI builder controls (models Check-in Config)

### PR DESCRIPTION
**Pull request by dcs619**
# Context
_What is the problem you encountered that lead to you creating this pull request?_

While beefing up Event Registration capabilities (future separate PR coming), I liked the UX speed and usability of the Check-in Config areas and wanted to follow the same functional pattern.  However, since the user would be able to add more generic Event resources, I need to be able to hide Check-in specific items.   I also wanted to control if/when resources could be added.

# Goal
_What will this pull request achieve and how will this fix the problem?_

Adds four new user controls that are a direct rip from Checkin Config:
- CheckinAreaRow -> ResourceAreaRow
- CheckinArea -> ResourceArea
- CheckinGroupRow -> ResourceGroupRow
- CheckinGroup -> ResourceGroup

Note: this functionality is not difficult to add directly to the CheckinXXX controls, but since it represents a usability shift I assumed it would be better to add completely new controls.

# Strategy
_How have you implemented your solution?_

I've copied the existing controls and renamed them (more generically) ResourceXXX, then added specific options for each control: 

ResourceArea.EnableCheckinOptions
- Will show or hide the `Check-in Labels` header and grid 
- Visible option in markup

ResourceAreaRow.EnableAddAreas
- Will enable or disable the Add button for new areas
- Not visible in markup, set from code-behind 

ResourceAreaRow.EnableAddGroups
- Will enable or disable the Add button for new groups
- Not visible in markup, set from code-behind 

ResourceAreaRow.btnDelete
- Visibility controlled from code-behind

ResourceGroup.EnableAddLocations
- Will show or hide the `Locations` header and grid
- Visible option in markup

ResourceGroupRow.EnableAddGroups
- Will enable or disable the Add button for new groups
- Not visible in markup, set from code-behind 

ResourceGroupRow.btnDelete
- Visibility controlled from code-behind

# Possible Implications
_What could this change potentially impact? Are there any security considerations? Where could this potentially affect backwards compatibility?_

N/A, additional functionality.

# Screenshots
_Provide us some screenshots if your pull request either alters existing UI or provides new UI. Arrows and labels are helpful._

### Before:

screen shot 2017-08-23 at 12 32 57 pm

screen shot 2017-08-23 at 12 33 32 pm

### After:

screen shot 2017-08-23 at 1 12 26 pm

screen shot 2017-08-23 at 1 12 46 pm

# Documentation
_If your change effects the UI or needs to be documented in one of the existing [user guides](http://www.rockrms.com/Learn/Documentation), please provide the brief write-up here:_

See above.

### CODE REVIEW

**Comment by edmistj** [comment]
We're going to hold on this until we see the big picture of what you're trying to accomplish.

File: Rock/Web/UI/Controls/Resource Controls/ResourceAreaRow.cs [hunk 1]
```diff
@@ -279,7 +279,11 @@ public override void RenderControl( HtmlTextWriter writer )
 
             _lbAddArea?.RenderControl( writer );
             writer.Write( " " );
-            _lbAddGroup?.RenderControl( writer );
```
Comment (edmistj) [inline Rock/Web/UI/Controls/Resource Controls/ResourceAreaRow.cs]: Our plan was to start using c# 6.0 features in the .dlls now. We can't use it in the web project yet as we'd need to wait until the hosting companies all support it.

**Comment by dcs619** [comment]
@edmistj My hope was to have the controls available to build on in v7 (without requiring a custom Rock.dll).  

The Registration block would be a larger PR but I assumed there would be back and forth discussions before it makes it into core (aka not going to hit v7).

**Comment by cabal95** [comment]
It looks like everything you have added is just web controls, which means you don't need a custom Rock.dll - just a new DLL that contains those controls.  I have done this a few times. Once the controls are implemented in Rock core you just switch the `<MyCustom:` to `<Rock:` and you're basically done.

**Pull request closed without merging**